WebSocketTests.testWebSocketClient test failure - incorrect server response

Log attached with full details of the failure:
>Test Case 'WebSocketTests.testWebSocketClient' started at 2021-05-06 15:35:09.572
[ INFO ] GET /ws [request-id: 4E8ECBAD-5745-4E92-B0B0-C8AAFFC0A155]
[ WARNING ] read(descriptor:pointer:size:): Connection reset by peer (errno: 104) [request-id: 4E8ECBAD-5745-4E92-B0B0-C8AAFFC0A155]
adding 1 to http_requests_total
adding 1 to http_request_errors_total
recording 601669829119 http_request_duration_seconds
/workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/Tests/VaporTests/WebSocketTests.swift:26: error: WebSocketTests.testWebSocketClient : XCTAssertEqual failed: ("internalServerError") is not equal to ("ok") - 
Fatal error: leaking promise created at (file: "/workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/Tests/VaporTests/WebSocketTests.swift", line: 10): file /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/Tests/VaporTests/WebSocketTests.swift, line 10
Current stack trace:
/workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/Tests/VaporTests/WebSocketTests.swift:27: error: WebSocketTests.testWebSocketClient : XCTAssertEqual failed: ("{"error":true,"reason":"read(descriptor:pointer:size:): Connection reset by peer (errno: 104)"}") is not equal to ("Hello, world!") - 
0    libswiftCore.so                    0x00007f4b366b7db0 swift_reportError + 50
1    libswiftCore.so                    0x00007f4b36729f60 _swift_stdlib_reportFatalErrorInFile + 115
2    libswiftCore.so                    0x00007f4b3643edfe <unavailable> + 1383934
3    libswiftCore.so                    0x00007f4b3643ea07 <unavailable> + 1382919
4    libswiftCore.so                    0x00007f4b3643efe8 <unavailable> + 1384424
5    libswiftCore.so                    0x00007f4b3643d2d0 _assertionFailure(_:_:file:line:flags:) + 520
6    vaporPackageTests.xctest           0x0000556d7e5c99c5 <unavailable> + 7109061
7    vaporPackageTests.xctest           0x0000556d7e6b16ae <unavailable> + 8058542
8    vaporPackageTests.xctest           0x0000556d7e6b1618 <unavailable> + 8058392
9    vaporPackageTests.xctest           0x0000556d7e5c93e4 <unavailable> + 7107556
10   vaporPackageTests.xctest           0x0000556d7e5c9bf5 <unavailable> + 7109621
11   libswiftCore.so                    0x00007f4b366b9d1b <unavailable> + 3984667
12   vaporPackageTests.xctest           0x0000556d7ed005f5 <unavailable> + 14673397
13   libswiftCore.so                    0x00007f4b366b9d1b <unavailable> + 3984667
14   vaporPackageTests.xctest           0x0000556d7e9a32a0 <unavailable> + 11145888
15   vaporPackageTests.xctest           0x0000556d7e9a3309 <unavailable> + 11145993
16   libswiftCore.so                    0x00007f4b366b9d1b <unavailable> + 3984667
17   libswiftCore.so                    0x00007f4b366b1210 swift_arrayDestroy + 59
18   libswiftCore.so                    0x00007f4b36460c90 _ContiguousArrayStorage.__deallocating_deinit + 39
0x556d7e0a5179, closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install() -> () at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/.build/checkouts/swift-backtrace/Sources/Backtrace/Backtrace.swift:67
0x556d7e0a5188, @objc closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install() -> () at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/<compiler-generated>:0
0x7f4b360e097f
0x7f4b3643d4e5
0x556d7e5c99c4, closure #1 () -> () in NIO.EventLoopFuture.deinit at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIO/EventLoopFuture.swift:423
0x556d7e6b16ad, closure #1 () -> Swift.Bool in implicit closure #1 () -> Swift.Bool in NIO.debugOnly(() -> ()) -> () at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIO/Utilities.swift:44
0x556d7e6b1617, NIO.debugOnly(() -> ()) -> () at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIO/Utilities.swift:44
0x556d7e5c93e3, NIO.EventLoopFuture.deinit at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIO/EventLoopFuture.swift:419
0x556d7e5c9bf4, NIO.EventLoopFuture.__deallocating_deinit at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIO/EventLoopFuture.swift:0
0x7f4b366b9d1a
0x556d7ed005f4, objectdestroy.5 at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/vapor/<compiler-generated>:0
0x7f4b366b9d1a
0x556d7e9a329f, NIOWebSocket.NIOWebSocketClientUpgrader.deinit at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIOWebSocket/NIOWebSocketClientUpgrader.swift:0
0x556d7e9a3308, NIOWebSocket.NIOWebSocketClientUpgrader.__deallocating_deinit at /workspace/.logs/**/*.log/00FB59CB-317C-4766-AF95-65F01B2A86C5/repos/swift-nio/Sources/NIOWebSocket/NIOWebSocketClientUpgrader.swift:0
0x7f4b366b9d1a
0x7f4b366b124a
0x7f4b36460cb6

--- a/Tests/VaporTests/WebSocketTests.swift
+++ b/Tests/VaporTests/WebSocketTests.swift
@@ -32,6 +32,9 @@ final class WebSocketTests: XCTestCase {
                 }
             }.flatMap {
                 return promise.futureResult
+            }.flatMapError { error in
+                promise.fail(error)
+                return promise.futureResult
             }
         }
 

--- a/Tests/VaporTests/WebSocketTests.swift
+++ b/Tests/VaporTests/WebSocketTests.swift
@@ -5,7 +5,7 @@ final class WebSocketTests: XCTestCase {
     func testWebSocketClient() throws {
         let server = Application(.testing)
 
-        server.http.server.configuration.port = 8086
+        server.http.server.configuration.port = 0
 
         server.webSocket("echo") { req, ws in
             ws.onText { ws.send($1) }
@@ -19,10 +19,16 @@ final class WebSocketTests: XCTestCase {
             server.shutdown()
         }
 
+        guard let localAddress = server.http.server.shared.localAddress,
+              let port = localAddress.port else {
+                  XCTFail("couldn't get port from \(app.http.server.shared.localAddress.debugDescription)")
+                  return
+              }
+
         app.get("ws") { req -> EventLoopFuture<String> in
             let promise = req.eventLoop.makePromise(of: String.self)
             return WebSocket.connect(
-                to: "ws://localhost:8086/echo",
+                to: "ws://localhost:\(port)/echo",
                 on: req.eventLoop
             ) { ws in
                 ws.send("Hello, world!")

--- a/Tests/VaporTests/WebSocketTests.swift
+++ b/Tests/VaporTests/WebSocketTests.swift
@@ -13,41 +13,34 @@ final class WebSocketTests: XCTestCase {
         server.environment.arguments = ["serve"]
         try server.start()
 
-        let app = Application(.testing)
         defer {
-            app.shutdown()
             server.shutdown()
         }
 
-        guard let localAddress = server.http.server.shared.localAddress,
-              let port = localAddress.port else {
-                  XCTFail("couldn't get port from \(app.http.server.shared.localAddress.debugDescription)")
-                  return
-              }
+        guard let localAddress = server.http.server.shared.localAddress, let port = localAddress.port else {
+            XCTFail("couldn't get port from \(server.http.server.shared.localAddress.debugDescription)")
+            return
+        }
 
-        app.get("ws") { req -> EventLoopFuture<String> in
-            let promise = req.eventLoop.makePromise(of: String.self)
-            return WebSocket.connect(
-                to: "ws://localhost:\(port)/echo",
-                on: req.eventLoop
-            ) { ws in
-                ws.send("Hello, world!")
-                ws.onText { ws, text in
-                    promise.succeed(text)
-                    ws.close().cascadeFailure(to: promise)
-                }
-            }.flatMap {
-                return promise.futureResult
-            }.flatMapError { error in
-                promise.fail(error)
-                return promise.futureResult
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+
+        let promise = elg.next().makePromise(of: String.self)
+        let string = try WebSocket.connect(
+            to: "ws://localhost:\(port)/echo",
+            on: elg.next()
+        ) { ws in
+            ws.send("Hello, world!")
+            ws.onText { ws, text in
+                promise.succeed(text)
+                ws.close().cascadeFailure(to: promise)
             }
-        }
-
-        try app.testable().test(.GET, "/ws") { res in
-            XCTAssertEqual(res.status, .ok)
-            XCTAssertEqual(res.body.string, "Hello, world!")
-        }
+        }.flatMap {
+            return promise.futureResult
+        }.flatMapError { error in
+            promise.fail(error)
+            return promise.futureResult
+        }.wait()
+        XCTAssertEqual(string, "Hello, world!")
     }
 
 

--- a/Tests/VaporTests/WebSocketTests.swift
+++ b/Tests/VaporTests/WebSocketTests.swift
@@ -3,13 +3,26 @@ import XCTest
 
 final class WebSocketTests: XCTestCase {
     func testWebSocketClient() throws {
+        let server = Application(.testing)
+
+        server.http.server.configuration.port = 8086
+
+        server.webSocket("echo") { req, ws in
+            ws.onText { ws.send($1) }
+        }
+        server.environment.arguments = ["serve"]
+        try server.start()
+
         let app = Application(.testing)
-        defer { app.shutdown() }
+        defer {
+            app.shutdown()
+            server.shutdown()
+        }
 
         app.get("ws") { req -> EventLoopFuture<String> in
             let promise = req.eventLoop.makePromise(of: String.self)
             return WebSocket.connect(
-                to: "ws://echo.websocket.org/",
+                to: "ws://localhost:8086/echo",
                 on: req.eventLoop
             ) { ws in
                 ws.send("Hello, world!")

